Constrained Beam Search Blog Post

cwkeam opened the pull request:
@patrickvonplaten Initial draft of a blog post to introduce & explain the new constrained beam search feature.

Written using a colab notebook found [here](https://github.com/cwkeam/blog/blob/cwkeam/constrained-beam-search/notebooks/19_constrained_beam_search.ipynb).

Not sure how the numberings worked, so I just added +1 to the highest number under `notebooks/`.

Since this feature isn't merged yet, the notebook temporarily installs my branch with

```
pip install -q git+https://github.com/cwkeam/transformers.git@constrained_beam_search
```

patrickvonplaten commented:
Very cool taking a look soon (will try this week) - at the very end we can convert it to markdown as well :-)

patrickvonplaten commented:
Overall - super cool blog post! Let me know once you're down and then I can help you convert the blog post to markdown and to put it live on https://huggingface.co/blog

cwkeam commented:
Thank you for the super detailed comments! I agree with all of them and have adjusted the post accordingly.

After looking through this repo I've done the following based on what I saw:

- Added a `thumbnail.png` and all diagrams are under `assets/54_constrained_beam_search`.
- added an entry under `_blog.yml`
- numbered my post as `54` even though the most recent in `master` is `52`, because it seems like there's a post that's about to be published about AI residency numbered at `53`.

And I'm happy with the post as it stands! How do we convert this into something like your post about text generation?

cwkeam commented:
@patrickvonplaten thanks for the pointers! I converted it to md and pushed it. I think it all worked as desired without any obvious problems, but it's hard to exactly imagine how it'd look like in deployment because the latex isnt rendered. Could you take a look as well?

patrickvonplaten commented on _blog.yml:
```diff
@@ -506,4 +506,13 @@
+  author: cwkeam
+  guest: true
+  thumbnail: /blog/assets/54_constrained_beam_search/thumbnail.jpg
+  date: March 7, 2022
```
```suggestion
  date: March 10, 2022
```

osanseviero commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+
+![Beam search](https://raw.githubusercontent.com/patrickvonplaten/scientific_images/master/beam_search.png)
+
+Unlike greedy search, beam search works by keeping a longer list of hypotheses. In the above picture, we have displayed 3 next possible tokens at each possible step in the generation.
```
```suggestion
Unlike greedy search, beam search works by keeping a long list of hypotheses. In the above picture, we have displayed three following possible tokens at each possible step in the generation.
```

cwkeam replied:
Tried to mean that greedy search just works with one hypotheses and beam search works with longer. Will change 3 -> three though

osanseviero commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+
+In the next step, we consider the next possible tokens for each of the three branches we created in the previous step.
+
+![Beam search step 2](https://raw.githubusercontent.com/cwkeam/scientific-images/main/beam_2.jpg)
```
Should this image and others be switched to the one in this repo?

cwkeam replied:
agreed. I was just worried one would get no images at all while reading this before merge

osanseviero commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+
+![Constrained Beam Search Step 3](https://raw.githubusercontent.com/cwkeam/scientific-images/main/cbeam_3.jpg)
+
+Notice how `"The is fast"` doesn't require any manual appending of constraint tokens since it's already fulfilled (i.e. already contains the phrase `"is fast"`). Also notice how beams like `"The dog is slow"` or `"The dog is mad"` is actually in Bank 0, since, although it includes the token `"is"`, it must restart from the beginning in order to generate `"is fast"`. By appending something like `" ...
```
```suggestion
Notice how `"The is fast"` doesn't require any manual appending of constraint tokens since it's already fulfilled (i.e., already contains the phrase `"is fast"`). Also, notice how beams like `"The dog is slow"` or `"The dog is mad"` is actually in Bank 0, since, although it includes the token `"is"`, it must restart from the beginning to generate `"is fast"`. By appending something like `"slow"` after `"is"`, it has effectively *reset its progress*. 
```

osanseviero commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+
+Notice how `"The is fast"` doesn't require any manual appending of constraint tokens since it's already fulfilled (i.e. already contains the phrase `"is fast"`). Also notice how beams like `"The dog is slow"` or `"The dog is mad"` is actually in Bank 0, since, although it includes the token `"is"`, it must restart from the beginning in order to generate `"is fast"`. By appending something like `" ...
+
+And finally notice how we ended up at a sensible output that contains our constraint phrase: `"The dog is fast"`! 
```
```suggestion
And finally, notice how we ended up at a sensible output that contains our constraint phrase: `"The dog is fast"`! 
```

osanseviero commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+
+And finally notice how we ended up at a sensible output that contains our constraint phrase: `"The dog is fast"`! 
+
+We were worried in the beginning because just blindly appending the desired tokens led to nonsensicle phrases like `"The is fast"`. However, with the use of round-robin selection from banks, we implicitly ended up getting rid of nonsensical outputs in preference for the more sensible outputs.  
```
```suggestion
We were worried initially because blindly appending the desired tokens led to nonsensical phrases like `"The is fast"`. However, using round-robin selection from banks, we implicitly ended up getting rid of nonsensical outputs in preference for the more sensible outputs. 
```

osanseviero commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+    Wie alt sind Sie?
+
+
+In order to design your own unique constraints, you can define one yourself and input it into the `constraints` keyword argument. You just have to create a sub-class of the `Constraint` abstract interface class and follow its requirements. More information about this can be found in the definition of `Constraint` found [here](https://github.com/huggingface/transformers/blob/master/src/transformer ...
```
```suggestion
You can define one yourself and input it into the `constraints` keyword argument to design your unique constraints. You just have to create a sub-class of the `Constraint` abstract interface class and follow its requirements. You can find more information in the definition of `Constraint` found [here](https://github.com/huggingface/transformers/blob/master/src/transformers/generation_beam_constraints.py).
```

osanseviero commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+
+## **Conclusion**
+
+Constrained beam search gives us a flexible means to inject external knowledge and requirements into text generation. Previously, there was no easy way to tell the model to 1. include a list of sequences where 2. some of which are optional and some are not, such that 3. they're generated *somewhere* in the sequence at reasonable respective positions. Now, with a mix of different subclasses of `Co ...
```
```suggestion
Constrained beam search gives us a flexible means to inject external knowledge and requirements into text generation. Previously, there was no easy way to tell the model to 1. include a list of sequences where 2. some of which are optional and some are not, such that 3. they're generated *somewhere* in the sequence at respective reasonable positions. Now, we can have full control over our generation with a mix of different subclasses of `Constraint` objects! 
```

osanseviero commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+ - [Improved Lexically Constrained Decoding for Translation and Monolingual Rewriting](https://aclanthology.org/N19-1090/)
+ - [Guided Generation of Cause and Effect](https://arxiv.org/pdf/2107.09846.pdf)
+
+Many new research papers recently, like the ones above, are exploring ways of using external knowledge (e.g. KGs, KBs) to guide the outputs of large deep learning models. Hopefully this constrained beam search feature becomes another effective way to achieve this purpose.
```
```suggestion
Like the ones above, many new research papers are exploring ways of using external knowledge (e.g., KGs, KBs) to guide the outputs of large deep learning models. Hopefully, this constrained beam search feature becomes another effective way to achieve this purpose.
```

osanseviero reviewed (approved):
This is a great blog post! I really liked it!

I left some minor suggestions for readibility, but feel free to disregard if you disagree with any. This is great content, thank you very much for contributing! 🚀 I learned new things!

cwkeam commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+
+### **Why It's Difficult**
+
+However, this is actually a very non-trivial problem. This is because the task requires us to force the generation of certain subsequences *somewhere* in the final output, at *some point* during the generation. 
```
[leaving this unchanged]
I'm saying "actually" since, on the surface, it looks like an easy problem to solve. The rest is personal writing style I think (think the shorter sentences are easier to read as well)

cwkeam commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+
+And what if you have multiple constraints with varying requirements? What if you want to force the phrase \\( p_1=\{t_1, t_2\} \\) *and* also the phrase \\( p_2=\{ t_3, t_4, t_5, t_6\} \\) ? What if you want the model to **choose between** the two phrases? What if we want to force the phrase \\( p_1 \\) and force just one phrase among the list of phrases \\( \{p_{21}, p_{22}, p_{23}\} \\) ? 
+
+The above are actually very reasonable use-cases, as it will be shown below, and the new constrained beam search feature allows for all of them!
```
[leaving unchanged]
im saying "actually" because the above example sounds very contrived, but they're actually very reasonable. Also it's not shown directly below but it's shown very much later, so current wording makes more sense I think (as shown below suggests its foud right below the paragraph).

cwkeam commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+    Wie alt sind Sie?
+
+
+As you can see, we were able to guide the generation with prior knowledge about our desired output. Previously we would've had to generate a bunch of possible outputs, then filter the ones that fit our requirement. Now we can do that at the generation stage.
```
[leaving unchanged] though I agree with taking out "..able to" for previous parts, I think there I want to emphasize our new ***ability*** to do this here. Also comma before "then" is fine I think and I like how it separates the two to convey the feeling that it's a more annoying and lengthier process

cwkeam commented on constrained-beam-search.md:
```diff
@@ -0,0 +1,362 @@
+        <img class="avatar avatar-user" src="https://avatars.githubusercontent.com/u/8522953?v=4" title="Gravatar">
+        <div class="bfc">
+            <code>cwkeam</code>
+            <span class="fullname">Chan Woo Kim</span>
```
done!

cwkeam commented:
@osanseviero thank you so much for the detailed feedback! I've adjusted the post accordingly. I've also

- changed the date to March 11, 2022
- changed it to # 53 since it seems like the AI residency post is coming later & theres another post coming up with number 54 

also, many of the links refer to would-be links on the master branch of this repo, which obviously don't work yet before merge. Is it possible to merge then see if there are no problems? I did double-check several times, but one can never be sure.

patrickvonplaten commented:
@osanseviero - I checked that locally the blog renders fine (only the images are not displayed, but should be once the PR is deployed)

osanseviero commented:
This is now deployed! :rocket: very nice job!

One item failing at the moment is the thumbnail https://huggingface.co/blog

Apart from that, the post looks great! https://huggingface.co/blog/constrained-beam-search
